--- /home/poid/opt-cm-18.1/bionic/libc/bionic/pthread_internal.cpp.orig	2025-08-17 19:09:34.383338253 +0900
+++ /home/poid/opt-cm-18.1/bionic/libc/bionic/pthread_internal.cpp	2025-08-21 22:10:34.488396597 +0900
@@ -99,7 +99,7 @@
       if (t == thread) return thread;
     }
   }
-
+#if 0 
   // Historically we'd return null, but from API level 26 we catch this error.
   if (android_get_application_target_sdk_version() >= 26) {
     if (thread == nullptr) {
@@ -114,5 +114,6 @@
 //      async_safe_fatal("invalid pthread_t %p passed to %s", thread, caller);
     }
   }
+#endif
   return nullptr;
 }
