diff --git ./bionic/libc/bionic/pthread_internal.cpp ./bionic/libc/bionic/pthread_internal.cpp.new
index 6fddefe43..637104f9b 100644
--- ./bionic/libc/bionic/pthread_internal.cpp
+++ ./bionic/libc/bionic/pthread_internal.cpp.new
@@ -101,6 +101,8 @@ pthread_internal_t* __pthread_internal_find(pthread_t thread_id, const char* cal
   }
 
   // Historically we'd return null, but
+  (void)caller;
+#if 0
   if (android_get_application_target_sdk_version() >= __ANDROID_API_O__) {
     if (thread == nullptr) {
       // This seems to be a common mistake, and it's relatively harmless because
@@ -113,5 +115,6 @@ pthread_internal_t* __pthread_internal_find(pthread_t thread_id, const char* cal
       async_safe_fatal("invalid pthread_t %p passed to %s", thread, caller);
     }
   }
+#endif
   return nullptr;
 }
diff --git ./bionic/linker/linker.cpp ./bionic/linker/linker.cpp.new
index 6941fde20..8c3e38d63 100644
--- ./bionic/linker/linker.cpp
+++ ./bionic/linker/linker.cpp.new
@@ -3973,6 +3973,17 @@ bool soinfo::link_image(const soinfo_list_t& global_group, const soinfo_list_t&
   if (has_text_relocations) {
     // Fail if app is targeting M or above.
     int app_target_api_level = get_application_target_sdk_version();
+#ifdef SDK_VERSION_OVERRIDES
+    for (const auto& entry : android::base::Split(SDK_VERSION_OVERRIDES, " ")) {
+      auto splitted = android::base::Split(entry, "=");
+      if (splitted.size() == 2 && splitted[0] == get_realpath()) {
+        app_target_api_level = static_cast<uint32_t>(std::stoul(splitted[1]));
+        DEBUG("\"%s\" has text relocations. Overriding sdk version %d to %d", get_realpath(), get_application_target_sdk_version(), app_target_api_level);
+        break;
+      }
+    }
+#endif
+
     if (app_target_api_level >= __ANDROID_API_M__) {
       DL_ERR_AND_LOG("\"%s\" has text relocations (https://android.googlesource.com/platform/"
                      "bionic/+/master/android-changes-for-ndk-developers.md#Text-Relocations-"
