diff -u -r ./frameworks/native/include/binder/Parcel.h ./frameworks/native/include/binder/Parcel.h.new
--- ./frameworks/native/include/binder/Parcel.h	2021-11-10 15:22:03.000000000 +0900
+++ ./frameworks/native/include/binder/Parcel.h.new	2021-12-24 08:27:48.186651445 +0900
@@ -577,9 +577,10 @@
     public:
         inline void* data() { return mData; }
     };
-
+#ifndef DISABLE_ASHMEM_TRACKING
 private:
     size_t mOpenAshmemSize;
+#endif
 
 public:
     // TODO: Remove once ABI can be changed.
diff -u -r ./frameworks/native/libs/binder/Android.bp ./frameworks/native/libs/binder/Android.bp.new
--- ./frameworks/native/libs/binder/Android.bp	2021-11-10 15:21:59.000000000 +0900
+++ ./frameworks/native/libs/binder/Android.bp.new	2021-12-24 08:22:28.978158965 +0900
@@ -108,6 +108,7 @@
         "-Wextra",
         "-Werror",
         "-Wzero-as-null-pointer-constant",
+	"-DDISABLE_ASHMEM_TRACKING",
     ],
     product_variables: {
         binder32bit: {
diff -u -r ./frameworks/native/libs/binder/Parcel.cpp ./frameworks/native/libs/binder/Parcel.cpp.new
--- ./frameworks/native/libs/binder/Parcel.cpp	2021-11-10 15:22:00.000000000 +0900
+++ ./frameworks/native/libs/binder/Parcel.cpp.new	2021-12-24 08:27:04.170699458 +0900
@@ -542,8 +542,11 @@
 
             flat_binder_object* flat
                 = reinterpret_cast<flat_binder_object*>(mData + off);
+#ifndef DISABLE_ASHMEM_TRACKING
             acquire_object(proc, *flat, this, &mOpenAshmemSize);
-
+#else
+            acquire_object(proc, *flat, this);
+#endif
             if (flat->hdr.type == BINDER_TYPE_FD) {
                 // If this is a file descriptor, we need to dup it so the
                 // new Parcel now owns its own fd, and can declare that we
@@ -1389,7 +1392,11 @@
         // Need to write meta-data?
         if (nullMetaData || val.binder != 0) {
             mObjects[mObjectsSize] = mDataPos;
+#ifndef DISABLE_ASHMEM_TRACKING
             acquire_object(ProcessState::self(), val, this, &mOpenAshmemSize);
+#else
+            acquire_object(ProcessState::self(), val, this);
+#endif
             mObjectsSize++;
         }
 
@@ -2630,7 +2637,11 @@
         i--;
         const flat_binder_object* flat
             = reinterpret_cast<flat_binder_object*>(data+objects[i]);
+#ifndef DISABLE_ASHMEM_TRACKING
         release_object(proc, *flat, this, &mOpenAshmemSize);
+#else
+        release_object(proc, *flat, this);
+#endif
     }
 }
 
@@ -2647,7 +2658,11 @@
         i--;
         const flat_binder_object* flat
             = reinterpret_cast<flat_binder_object*>(data+objects[i]);
+#ifndef DISABLE_ASHMEM_TRACKING
         acquire_object(proc, *flat, this, &mOpenAshmemSize);
+#else
+        acquire_object(proc, *flat, this);
+#endif
     }
 }
 
@@ -2842,7 +2857,11 @@
                     // will need to rescan because we may have lopped off the only FDs
                     mFdsKnown = false;
                 }
+#ifndef DISABLE_ASHMEM_TRACKING
                 release_object(proc, *flat, this, &mOpenAshmemSize);
+#else
+                release_object(proc, *flat, this);
+#endif
             }
 
             if (objectsSize == 0) {
@@ -2937,7 +2956,9 @@
     mFdsKnown = true;
     mAllowFds = true;
     mOwner = nullptr;
+#ifndef DISABLE_ASHMEM_TRACKING
     mOpenAshmemSize = 0;
+#endif
     mWorkSourceRequestHeaderPosition = 0;
     mRequestHeaderPresent = false;
 
@@ -2974,12 +2995,20 @@
     // This used to return the size of all blobs that were written to ashmem, now we're returning
     // the ashmem currently referenced by this Parcel, which should be equivalent.
     // TODO: Remove method once ABI can be changed.
+#ifndef DISABLE_ASHMEM_TRACKING
     return mOpenAshmemSize;
+#else
+    return 0;
+#endif	
 }
 
 size_t Parcel::getOpenAshmemSize() const
 {
+#ifndef DISABLE_ASHMEM_TRACKING
     return mOpenAshmemSize;
+#else
+    return 0;
+#endif	
 }
 
 // --- Parcel::Blob ---
diff -u -r ./frameworks/native/libs/binder/include/binder/Parcel.h ./frameworks/native/libs/binder/include/binder/Parcel.h.new
--- ./frameworks/native/libs/binder/include/binder/Parcel.h	2021-11-10 15:22:03.000000000 +0900
+++ ./frameworks/native/libs/binder/include/binder/Parcel.h.new	2021-12-24 08:27:48.186651445 +0900
@@ -577,9 +577,10 @@
     public:
         inline void* data() { return mData; }
     };
-
+#ifndef DISABLE_ASHMEM_TRACKING
 private:
     size_t mOpenAshmemSize;
+#endif
 
 public:
     // TODO: Remove once ABI can be changed.

