diff --git ./system/core/adb/daemon/reboot_service.cpp ./system/core/adb/daemon/reboot_service.cpp
index a5a11b866..8937123dd 100644
--- ./system/core/adb/daemon/reboot_service.cpp
+++ ./system/core/adb/daemon/reboot_service.cpp.new
@@ -42,7 +42,8 @@ void reboot_service(unique_fd fd, const std::string& arg) {
     std::string reboot_string = android::base::StringPrintf("reboot,%s", reboot_arg.c_str());
 
     if (reboot_arg == "fastboot" &&
-        android::base::GetBoolProperty("ro.boot.dynamic_partitions", false) &&
+        (android::base::GetBoolProperty("ro.boot.dynamic_partitions", false) ||
+         android::base::GetBoolProperty("ro.fastbootd.available", false)) &&
         access("/dev/socket/recovery", F_OK) == 0) {
         LOG(INFO) << "Recovery specific reboot fastboot";
         /*
diff --git ./system/core/init/reboot_utils.cpp ./system/core/init/reboot_utils.cpp
index de085cc9b..36d39b50d 100644
--- ./system/core/init/reboot_utils.cpp
+++ ./system/core/init/reboot_utils.cpp
@@ -33,7 +33,7 @@
 namespace android {
 namespace init {
 
-static std::string init_fatal_reboot_target = "bootloader";
+static std::string init_fatal_reboot_target = "recovery";
 
 void SetFatalRebootTarget() {
     std::string cmdline;
diff --git ./system/core/init/reboot.cpp ./system/core/init/reboot.cpp.new
--- ./system/core/init/reboot.cpp	2025-06-09 21:43:59.601661244 +0900
+++ ./system/core/init/reboot.cpp.new	2025-06-09 21:45:46.969130887 +0900
@@ -544,11 +544,11 @@
             if (reboot_target == "fastboot" &&
                 !android::base::GetBoolProperty("ro.boot.dynamic_partitions", false) &&
                 !android::base::GetBoolProperty("ro.fastbootd.available", false)) {
-                reboot_target = "bootloader";
+                reboot_target = "recovery";
             }
             // When rebooting to the bootloader notify the bootloader writing
             // also the BCB.
-            if (reboot_target == "bootloader") {
+            if (reboot_target == "recovery") {
                 std::string err;
                 if (!write_reboot_bootloader(&err)) {
                     LOG(ERROR) << "reboot-bootloader: Error writing "
